### ISSUE TITLE
Wrong platform assets being copied to flutter_assets, macOS assets showing up in iOS flutter_assets

### ISSUE THREAD

**Issue by AfzalivE** [flutter/flutter#45075]
## Steps to Reproduce

1. Setup macos support in an existing project with Android and iOS support. `flutter config --enable-macos-desktop` and then `flutter create --macos .` (or whatever you need to do)
2. Build the release version of the Mac part `flutter build macos`
3. Build the release version of the iOS part `flutter build ios`

Maybe try a few more times in that order if everything works fine the first time. 

**Target Platform: Mac and iOS**
**Target OS version/browser: Mac OS 10.15.1 and iOS 13**

## Logs

```
 === BUILD TARGET Runner OF PROJECT Runner WITH CONFIGURATION Release ===
    2019-11-17 15:05:59.049 defaults[4938:4863527]
    The domain/default pair of
    (/Users/afzal/Dev/projects/Flutter/mailgun_routes/build/ios/Release-iphoneos/Ma
    ilgun
    Routes.app/Frameworks/App.framework/flutter_assets/macos/Flutter/ephemeral/App.
    framework/Info.plist, CFBundleExecutable) does not exist
    fatal error: lipo: can't map input file:
    /Users/afzal/Dev/projects/Flutter/mailgun_routes/build/ios/Release-iphoneos/Mai
    lgun
    Routes.app/Frameworks/App.framework/flutter_assets/macos/Flutter/ephemeral/App.
    framework/ (Invalid argument)
    fatal error: lipo: can't map input file:
    /Users/afzal/Dev/projects/Flutter/mailgun_routes/build/ios/Release-iphoneos/Mai
    lgun
    Routes.app/Frameworks/App.framework/flutter_assets/macos/Flutter/ephemeral/App.
    framework/ (Invalid argument)
    Failed to extract armv7 for
    /Users/afzal/Dev/projects/Flutter/mailgun_routes/build/ios/Release-iphoneos/Mai
    lgun
    Routes.app/Frameworks/App.framework/flutter_assets/macos/Flutter/ephemeral/App.
    framework/. Running lipo -info:
    fatal error: lipo: can't map input file:
    /Users/afzal/Dev/projects/Flutter/mailgun_routes/build/ios/Release-iphoneos/Mai
    lgun
    Routes.app/Frameworks/App.framework/flutter_assets/macos/Flutter/ephemeral/App.
    framework/ (Invalid argument)
```

For some reason, the **macos** folder gets created in `ios/Flutter/App.Framework/flutter_assets` when building the iOS version after the Mac version.

This also continues happening if you add the web part to this project. The `build/web/assets` folder will contain `ios` and `macos` folders with `Flutter/App.Framework` inside them (and flutter_assets) inside those. I think something in the build cache might be causing this.

Note: I am importing a json asset in pubspec.yaml called "secrets.json"
This is the project:
https://github.com/AfzalivE/mailgun_routes

## Workaround

https://github.com/flutter/flutter/issues/45075#issuecomment-554811637

**Comment by AfzalivE** [comment]
If pointed in the right direction, I can also try to fix this in flutter tools

**Comment by AfzalivE** [comment]
It seems like this is happening because of _parseAssetsFromFolder in asset.dart goes through the macos folder and finds the assets again, because they're present in `/Users/afzal/Dev/projects/Flutter/mailgun_routes/macos/Flutter/ephemeral/App.framework/Resources/flutter_assets/`

This is incorrect. Maybe App.Framework should be in the list of excludedDirs for a start?

**Comment by AfzalivE** [comment]
Alright so the fix for my project is to simply create an assets folder in the root of my project and move secrets.json to it.

That seems to fix the crazy stuff where the `macos/Flutter/ephemeral/App.framework/Resources/flutter_assets` gets included in the ios app and causes the lipo error.

Whew. That was a deep dive into the flutter build system to find this fix :/ 

There should be some documentation informing people that if they put their assets outside of the assets folder, it's not going to be a good 7 hours for them.

**Comment by jmagman** [comment]
Wow @AfzalivE thanks for all that digging!  It sounds like you did 95% of the work, if you're interested in sending us a pull request to exclude the ephemeral directory from that search (plus a new test in asset_test.dart).  😉 
https://github.com/flutter/flutter/wiki/The-flutter-tool#making-changes-to-the-flutter-tool

**Comment by stuartmorgan** [comment]
I'm not familiar with that part of the code, but is there any chance the search can be whitelist-based rather than blacklist? In a future where we support extensible platform definitions, we won't reliably be able to blacklist everything.

**Comment by jonahwilliams** [comment]
I think this the same issue as https://github.com/flutter/flutter/issues/45100

**Comment by jmagman** [comment]
> I created a demo for this bug: https://github.com/jld3103/flutter_assets_bug
> 
> When you use `flutter build web` / `flutter build bundle` files which aren't in the assets, but have the same name. The are copied to `build/web/assets` / `build/flutter_assets/`.
> 
> This needs to be fixed very soon.

@jld3103 made a demo app at https://github.com/flutter/flutter/issues/45100.

@jonahwilliams from https://github.com/flutter/flutter/issues/45100#issuecomment-555258421:
> So the issue here is that this code is triggering our variant detection logic. Normally variants are handled such that all of the following assets are bundled with a single `a.png` asset listed in the manifest.
> 
> ```
> 2x/a.png
> 1x/a.png
> a.png
> ```
> 
> There isn't any restriction on what the variant should be called though, which leads to us picking up the random txt file.
> 
> For reference:
> 
> https://github.com/flutter/flutter/blob/ad1d67e0a1ebc444ec9e89573d6df97a2fdb2e41/packages/flutter_tools/lib/src/asset.dart#L499

**Comment by jonahwilliams** [comment]
> I'm not familiar with that part of the code, but is there any chance the search can be whitelist-based rather than blacklist? 

When the user specifies the project root as the asset directory, I think a blacklist is the best we can do. Variants are currently allowed to be arbitrary strings.

- In the short term, we should add the remaining platforms to the excludeDirs (this won't fix  #45100) to mitigate the issue.
- In the long-term, we could either restrict or rethink variants (allow certain patterns only) or enforce that assets must be located in a subdirectory. Both would be breaking changes, but the former might not be so bad if we detect + warn for a period of time.

**Comment by AfzalivE** [comment]
I haven't been able to write a decent unit test in `assets_test.dart` because this pubspec.yaml (in the unit test) actually fails to create an `AssetManifest.json`:

```
name: example

flutter:
  assets:
    - hello.json
```

**Comment by baylessj** [comment]
@AfzalivE Thank you so much for finding this workaround! I also got stuck with about 7 hours of broken builds and frustration before finding your fix. 

This definitely should be somewhere in the docs, it was a major headache to debug and this issue was much harder to find on google with the flutter output logs than I would have expected.

**Comment by elanthi** [comment]
I wasted 1 day trying to fix this issue. I even went on Flutter 1.21.0-1.0.pre • channel dev, but nothing worked.
Later I fixed it by following these steps 
Go to XCode-> Preferences-> Locations-> Derived Data 
Open Derived data folder and delete everything inside it
flutter clean
pod install
Build in Xcode And then everything worked as expected.

Also be careful to not have project inside a folder with the same name as the project like /projectName/projectName/...
(I had it like this, but I dont know if it was relevant, maybe the derived data deletion itself would have worked the same)
